File: 2026-09/day0068/f1/static/app1.js
const newTitleInput = document.querySelector("#new_title");
const newCategoryInput = document.querySelector("#new_category");
const newReadDateInput = document.querySelector("#new_read_date");
const newPagesInput = document.querySelector("#new_pages");

function addBook() {
    const new_title = newTitleInput.value;
    const new_category = newCategoryInput.value;
    const new_read_date = newReadDateInput.value;
    const new_pages = newPagesInput.value;

    fetch("/books", {
        method: "POST",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify(
            {
                "new_title": new_title,
                "new_category": new_category,
                "new_read_date": new_read_date,
                "new_pages": new_pages
            })
    })
        .then(response => {
            if (!response.ok) {
                return response.json()
                    .then(error => {
                        throw new Error(error.result);
                    });
            }

            return response.json();
        })
        .then(data => {
            loadBooks();
        })
        .catch(error => {
            console.log(error.message);
        });
}

const addButton = document.querySelector("#add-button")
addButton.addEventListener("click", addBook)


const categorySelect = document.querySelector("#category")
const startDateSelect = document.querySelector("#start_date")
const endDateSelect = document.querySelector("#end_date")
const sortSelect = document.querySelector("#sort")

categorySelect.addEventListener("change", loadBooks)
startDateSelect.addEventListener("change", loadBooks)
endDateSelect.addEventListener("change", loadBooks)
sortSelect.addEventListener("change", loadBooks)

function loadBooks() {
    const category = categorySelect.value;
    const start_date = startDateSelect.value;
    const end_date = endDateSelect.value;
    const sort = sortSelect.value;

    let queryParams = new URLSearchParams();

    if (category) {
        queryParams.append("category", category);
    }

    if (start_date) {
        queryParams.append("start_date", start_date)
    }

    if (end_date) {
        queryParams.append("end_date", end_date)
    }

    if (sort) {
        queryParams.append("sort", sort)
    }

    const queryString = queryParams.toString();

    let url = "/books"

    if (queryString) {
        url = url + "?" + queryString
    }

    fetch(url)
        .then(response => {
            if (!response.ok) {
                return response.json()
                    .then(error => {
                        throw new Error(error.result)
                    })
            }

            return response.json()
        })
        .then(data => {
            const books = data.books_list;
            const booksList = document.querySelector("#books-list")

            booksList.innerHTML = ""

            books.forEach(book => {
                const div = document.createElement("div")

                const h2 = document.createElement("h2")
                const li1 = document.createElement("li")
                const li2 = document.createElement("li")

                const updateButton = document.createElement("button")
                const deleteButton = document.createElement("button")


                h2.textContent = `${book.title}`
                li1.textContent = `${book.category}`
                li2.textContent = `읽은 날짜${book.read_date}, 읽은 페이지 수${book.pages}`


                updateButton.textContent = "변경"
                deleteButton.textContent = "삭제"


                updateButton.addEventListener("click", () => {
                    updateBook(`${book._id}`)
                })
                deleteButton.addEventListener("click", () => {
                    deleteBook(`${book._id}`)
                })

                div.appendChild(h2)
                div.appendChild(li1)
                div.appendChild(li2)
                div.appendChild(updateButton)
                div.appendChild(deleteButton)

                booksList.appendChild(div)
            })
        })
        .catch(error => {
            console.log(error.message);
        });
}

loadBooks();

function updateBook(book_id) {
    const update_title = document.querySelector("#update_title").value;
    const update_category = document.querySelector("#update_category").value;
    const update_read_date = document.querySelector("#update_read_date").value;
    const update_pages = document.querySelector("#update_pages").value;

    const updateData = {};

    if (update_title !== "") {
        updateData["update_title"] = update_title;
    }

    if (update_category !== "") {
        updateData["update_category"] = update_category;
    }

    if (update_read_date !== "") {
        updateData["update_read_date"] = update_read_date;
    }

    if (update_pages !== "") {
        updateData["update_pages"] = Number(update_pages);
    }

    fetch(`/books/${book_id}`, {
        method: "PATCH",
        headers: {
            "Content-Type": "application/json"
        },
        body: JSON.stringify({
            "update_data": updateData
        })
    })
        .then(response => {
            if (!response.ok) {
                return response.json()
                    .then(error => {
                        throw new Error(error.error);
                    });
            }

            return response.json();
        })
        .then(data => {
            loadBooks();
        })
        .catch(error => {
            console.log(error.message);
        });
}

function deleteBook(book_id) {
    fetch(`/books/${book_id}`, {
        method: "DELETE",
    })
        .then(response => {
            if (!response.ok) {
                return response.json()
                    .then(error => {
                        throw new Error(error.result);
                    });
            }

            return response.json();
        })
        .then(data => {
            loadBooks();
        })
        .catch(error => {
            console.log(error.message);
        });
}

File: 2026-09/day0068/f1/app1.py
from flask import Flask, request, render_template,jsonify
from pymongo import MongoClient
from bson import ObjectId

app = Flask(__name__)

client = MongoClient("mongodb://localhost:27017")
db = client["test0910"]
books = db["books"]

allowed_category = ["programming", "novel", "essay"]
allowed_sort = ["latest", "oldest"]


@app.route("/")
def home():
    return render_template("app1.html")

@app.route("/books", methods=["GET", "POST"])
def find_add_books():
    if request.method == "POST":
        data = request.get_json()

        new_title = data.get("new_title")
        new_category = data.get("new_category")
        new_read_date = data.get("new_read_date")
        new_pages = data.get("new_pages")

        if new_category and new_category not in allowed_category:
            return jsonify(
                {"result": "invalid category"}
            ), 400

        books.insert_one(
            {"title": new_title,
             "category": new_category,
             "read_date": new_read_date,
             "pages": new_pages}
        )

        return jsonify(
            {"result": "add success"}
        )

    elif request.method == "GET":
        category = request.args.get("category")
        start_date = request.args.get("start_date")
        end_date = request.args.get("end_date")
        sort = request.args.get("sort")

        query = {}

        if category:
            if category not in allowed_category:
                return jsonify(
                    {"result": "invalid category"}
                    ), 400
            query["category"] = category

        if sort and sort not in allowed_sort:
            return jsonify(
                {"result": "invalid sort"}
                ), 400

        query_date = {}

        if start_date and end_date and start_date > end_date:
            return jsonify(
                {"result": "invalid read_date"}
                ), 400

        if start_date:
            query_date["$gte"] = start_date

        if end_date:
            query_date["$lte"] = end_date

        if query_date:
            query["read_date"] = query_date

        sort_direction = -1

        if sort == "oldest":
            sort_direction = 1

        books_list = list(books.find(query).sort("read_date", sort_direction))

        for book in books_list:
            book["_id"] = str(book["_id"])

        return jsonify(
            {"books_list": books_list}
        )


@app.route("/books/<book_id>", methods=["PATCH"])
def update_book(book_id):
    update_datas = request.get_json()
    update_data = update_datas.get("update_data")

    update_title = update_data.get("update_title")
    update_category = update_data.get("update_category")
    update_read_date = update_data.get("update_read_date")
    update_pages = update_data.get("update_pages")

    update_query = {}

    if "update_title" in update_data:
            update_query["title"] = update_title

    if "update_category" in update_data:
            if update_category not in allowed_category:
                 return jsonify({
                    "error": "invalid category"
                })
            update_query["category"] = update_category

    if "update_read_date" in update_data:
            update_query["read_date"] = update_read_date
    
    if "update_pages" in update_data:
            update_query["pages"] = update_pages

    if not update_query:
        return jsonify({"error": "invalid query"}), 400

    try:
        book_id = ObjectId(book_id)
    except:
        return jsonify({
            "error": "잘못된 id 형식입니다."
        }), 400

    update_result = books.update_one(
        {"_id": book_id},
        {"$set": update_query}
    )

    if update_result.matched_count == 0:
        return jsonify({
            "error": "기록을 찾을 수 없습니다."
        }), 404

    return jsonify({
            "result": "update success"
        })

@app.route("/books/<book_id>", methods=["DELETE"])
def delete_book(book_id):
    books.delete_one(
        {"_id": ObjectId(book_id)},
    )

    return jsonify(
            {"result": "delete success"}
        )



if __name__ == "__main__":
    app.run(debug=True)


